modules/engine-prototype/src/main/scala/engine/Step.scala

```diff
@@ -0,0 +1,251 @@
+                          s.core,
+                          tcsFailed = true,
+                          instFailed = true,
+                          observationFailed = false
+                        )
+                      )
+                      // TCS didn't fail yet
+                      case s => Failed.Standard(
+                        F2.Failed.Standard(
+                          s.core,
+                          tcsFailed = false,
+                          instFailed = true,
+                          observationFailed = false
+                        )
+                      )
+                    }
+                  ) *> false.pure[F]
+              }
+
+              // XXX: Check running status
+              both(configureTCS, configureInst).flatMap {
+              case (true, true) =>
+                  // XXX: Check running status
+                  Execution.observe.flatMap {
+                    case Some(fileId) => execute(m) // XXX: Shift current focus
+                    case None =>
+                      m.modify(
+                        Sequence.State.current.modify(s =>
+                          Failed.Standard(
+                            F2.Failed.Standard(
+                              s.core,
+                              tcsFailed = false,
+                              instFailed = false,
+                              observationFailed = true
+                            )
+                          )
+                        )
+                      ).void
+                  }
+
+              case (_, _) => F.pure(Unit)
+            }
+          }
+        }
+      }
+
+      sealed trait Ongoing extends Current
+      object Ongoing {
+        case class Standard(step: F2.Ongoing.Standard) extends Ongoing
+      }
+
+      sealed trait Failed extends Current
+      object Failed {
+        case class Standard(step: F2.Failed.Standard) extends Failed
+      }
+    }
+
+  }
+
+  sealed trait GMOS extends Step
+  case object GMOS
+
+  // This won't be accepted in cats-effect
+  // https://gist.github.com/djspiewak/a775b73804c581f4028fea2e98482b3c but with
+  // fs2 it should be possible, `parallelSequence` is already there.
+  private def both[F[_], A, B](fa: F[A], fb: F[B])(implicit F: Effect[F]): F[(A, B)] = ???
+
+}
+
+// current match {
+//   // No current Step
+//   case None => pending match {
+//     // No pending steps, done
+//     case Nil => m.setStatus(Status.Finished).void
+//     // More pending steps
+//     case (next :: remainder) => next.execute(m).flatMap {
+//       case Left(e) => m.setState(
+//         Sequence.State(
+//           F2(done, remainder, Some(Left(e))),
+//           Status.Failed
+//         )
+//       )
+//       case Right(d) => m.setSequence(
+//         F2((d :: done), remainder, None)
+//       ).flatMap(_.sequence.execute(m))
+//     }
+//   }
+//   case Some(_) => F.pure(Unit) // Event: Tried execute Sequence with an already ongoing step.
+// }
+// }
```

Comment:
This is an early attempt with Current Steps where I'm doing some State shifting. I didn't delete it because I was going to recycle the shifting part in the new code.